flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java

```diff
@@ -563,18 +611,53 @@ private void addEdgeInternal(Integer upStreamVertexID,
 				}
 			}
 
-			if (shuffleMode == null) {
-				shuffleMode = ShuffleMode.UNDEFINED;
+			if (shuffleMode == null || shuffleMode == ShuffleMode.UNDEFINED) {
+				shuffleMode = determineShuffleMode(partitioner);
 			}
 
-			StreamEdge edge = new StreamEdge(upstreamNode, downstreamNode, typeNumber, outputNames, partitioner, outputTag, shuffleMode);
+			StreamEdge edge = new StreamEdge(
+				upstreamNode,
+				downstreamNode,
+				typeNumber,
+				outputNames,
+				partitioner,
+				outputTag,
+				shuffleMode);
 
 			getStreamNode(edge.getSourceId()).addOutEdge(edge);
 			getStreamNode(edge.getTargetId()).addInEdge(edge);
 		}
 	}
 
-	public <T> void addOutputSelector(Integer vertexID, OutputSelector<T> outputSelector) {
+	private ShuffleMode determineShuffleMode(StreamPartitioner<?> partitioner) {
+		switch (globalDataExchangeMode) {
+			case ALL_EDGES_BLOCKING:
+				return ShuffleMode.BATCH;
+			case FORWARD_EDGES_PIPELINED:
+				if (partitioner instanceof ForwardPartitioner) {
+					return ShuffleMode.PIPELINED;
+				} else {
+					return ShuffleMode.BATCH;
+				}
+			case POINTWISE_EDGES_PIPELINED:
+				if (isPointwisePartitioner(partitioner)) {
+					return ShuffleMode.PIPELINED;
+				} else {
+					return ShuffleMode.BATCH;
+				}
+			case ALL_EDGES_PIPELINED:
```

Comment:
I would put this check as second as it is one of the "straight-forward" ones. But this is just personal taste.